### PR TITLE
feat: add support for vendor extensions

### PR DESCRIPTION
**Pull request by weyert**
## Motivation and Context
Adds support for rendering vendor extensions as part of a schema row or the top level schema

Fixes https://github.com/stoplightio/json-schema-viewer/issues/236

## Description
Introduces support to custom renderer for vendor extensions, by adding a new property named `renderExtensionAddon` which will call when vendor properties (properties prefixed with `x-`) are find the schema definition.

## How Has This Been Tested?
Tested in production for our internal documentation

## Screenshot(s)/recordings(s)
Rendering of a vendor extension that shows the `x-stoplight` extension, in this example, it only returns a React element for any extensions with a nesting level higher than 0
CleanShot 2024-02-20 at 15 28 52@2x

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [X] This PR's code follows as closely as possible [the coding style/guidelines of this project](???).
- [ ] I have added error reporting and followed [the error reporting guidelines](???).
- [ ] I have added event tracking and followed [the event tracking guidelines](???).
- [ ] I have updated any relevant documentation accordingly to reflect this PR's changes.
- [X] I have added automated tests (unit/integration/e2e/other) to cover my changes.
- [X] All new and existing tests pass locally (excluding flaky CI tests).

### CODE REVIEW

**Comment by weyert** [comment]
@billiegoose @daniel-white @brendarearden  If someone could have a look at this would be great!

**Comment by weyert** [comment]
@P0lip Can you have a look?

File: src/components/SchemaRow/SchemaRow.tsx [hunk 1]
```diff
@@ -65,6 +67,12 @@ export const SchemaRow: React.FunctionComponent<SchemaRowProps> = React.memo(
     const validations = isRegularNode(schemaNode) ? schemaNode.validations : {};
     const hasProperties = useHasProperties({ required, deprecated, validations });
 
+    const [totalVendorExtensions, vendorExtensions] = React.useMemo(
+      () => extractVendorExtensions(schemaNode.fragment),
+      [schemaNode],
```
Comment (P0lip) [inline src/components/SchemaRow/SchemaRow.tsx:72]:
```suggestion
      [schemaNode.fragment],
```

File: src/utils/extractVendorExtensions.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+export type VendorExtensionsResult = [number, VendorExtensionsList];
+
+/**
+ * Extract all vendor extensions or properties prefix with 'x-' from the schema definition
+ * @param fragment The fragment to extract the vendor extensions from
+ * @returns VendorExtensionsResult
+ */
+export function extractVendorExtensions(fragment: SchemaFragment = {}): VendorExtensionsResult {
+  const extensionKeys = Object.keys(fragment).filter(key => key.toLowerCase().startsWith('x-'));
```
Comment (P0lip) [inline src/utils/extractVendorExtensions.ts:15]: I'm not sure we should lower-case the key. If a key starts with an upper-cased X-, then IMHO it shouldn't be considered a spec ext.

File: src/components/SchemaRow/TopLevelSchemaRow.tsx [hunk 3]
```diff
@@ -18,18 +20,28 @@ export const TopLevelSchemaRow = ({
...
+  const { renderExtensionAddon } = useJSVOptionsContext();
+
   const { selectedChoice, setSelectedChoice, choices } = useChoices(schemaNode);
   const childNodes = React.useMemo(() => visibleChildren(selectedChoice.type), [selectedChoice.type]);
   const nestingLevel = 0;
 
   const nodeId = schemaNode.fragment?.['x-stoplight']?.id;
+  const [totalVendorExtensions, vendorExtensions] = React.useMemo(
+    () => extractVendorExtensions(schemaNode.fragment),
+    [schemaNode],
```
Comment (P0lip) [inline src/components/SchemaRow/TopLevelSchemaRow.tsx:32]: ditto. Maybe extract it to a hook?

File: src/__fixtures__/real-world/vendor-extensions.json [hunk 4]
```diff
@@ -0,0 +1,517 @@
+{
```
Comment (P0lip) [inline src/__fixtures__/real-world/vendor-extensions.json:1]: Could you make this example a bit smaller? Could be just a few lines of schema to showcase extensions.

**Review by P0lip — COMMENTED** [review]
Looks good overall! Thanks for the PR.
I'd appreciate it if you added a test - LMK if you have a question regarding that. Happy to help if needed

**Comment by weyert** [comment]
Thanks, I will update the PR ASAP. I need to solve some work tasks first

File: src/__fixtures__/real-world/vendor-extensions.json (hunk 4 above)
Comment (weyert) [reply src/__fixtures__/real-world/vendor-extensions.json]: I have changed it and split it out of `real-world` its own section

**Comment by weyert** [comment]
Anymore changes needed?  I would love to be able to stop maintaining a fork of `elements` and `json-schema-viewer`
cc @P0lip @billiegoose @daniel-white @brendarearden @kaylachun

**Review by P0lip — APPROVED** [review]
Thank you!

**Comment by P0lip** [comment]
I'll merge it tomorrow.
Apologies for dropping a ball - I had time off last week too.

**Comment by weyert** [comment]
Awesome excited 👯‍♀️ :) Thank you! 🙏

**Comment by weyert** [comment]
@P0lip Would you be able to merge it? Love to raise a PR in Elements to bump with this support :)

**Comment by weyert** [comment]
Awesome thank you 👯

**Pull request merged**